docs/proposals/20191030-machine-health-checking.md

```diff
@@ -0,0 +1,211 @@
+
+It provides a short-circuit mechanism and limits remediation when the `maxUnhealthy` threshold is reached for a targeted group of machines. This is similar to what the node life cycle controller does for reducing the eviction rate as nodes become unhealthy in a given zone. E.g a large number of nodes in a single zone are down due to a networking issue.
+
+The machine health checker is an integration point between node problem detection tooling expresed as node conditions and remediation to achieve a node auto repairing feature.
+
+### Unhealthy criteria:
+A machine is unhealthy when:
+- The referenced node meets the unhealthy node conditions criteria defined.
+- The Machine has no nodeRef.
+- The Machine has a nodeRef but the referenced node is not found.
+- The Machine is in phase "Failed", It has `ErrorMessage`/`ErrorReason` set.
+
+If any of those criteria are met for longer than the given timeouts and the `maxUnhealthy` threshold has not been reached yet, remediation is triggered.
+
+Timeouts:
+- For the node conditions the time outs are defined by the admin.
+- For a machine with no nodeRef an opinionated value could be assumed e.g 10 min.
+- For a node notFound or a failed machine, the machine is considerable unrecoverable, remediation can be triggered right away if `maxUnhealthy` is not reached.
+
+### Remediation:
+- A deletion request for the machine is sent to the API server.
+- The controller owning that `Machine`, e.g machineSet reconciles towards the expected number of replicas and start the process to bring up a new `Machine`.
+- The machine controller drains the node.
+- The machine controller and the provider implementation delete the cloud instance.
+- The machine controller removes the finalizer for the `Machine` resource.
+
+
+### User Stories
+
+#### Story 1
+- As a user of a Workload Cluster I only care about my apps availability so I want my cluster infrastructure to be self healing and the nodes to be remediated transparently in the case of failures.
+
+#### Story 2
+- As an operator of a Management Cluster I want my machines to be self-healing and to be recreated resulting in a new healthy node in the case of matching my unhealthy criteria.
+
+
+### Implementation Details/Notes/Constraints
+
+#### MachineHealthCheck CRD:
+- Enable watching a group of machines (based on a label selector).
+- Enable defining an unhealthy node criteria (based on a list of node conditions).
+- Enable setting a threshold of unhealthy nodes. If the current number is at or above this threshold no further remediation will take place. This can be expressed as an int or as a percentage of the total targets in the pool.
+
+
+E.g:
+- I want my worker machines to be remediated when the backed node has `ready=false` or `ready=Unknown` condition for more than 5m.
+- I want remediation to temporary short-circuit if the 40% or more of the targets of this pool are unhealthy at the same time.
+
+```yaml
+apiVersion: healthchecking.cluster-api.io/v1alpha3
```

Comment:
I think @timothysc wants us to try to move toward a single API group for everything going forward: cluster.x-k8s.io.